madgraph/various/Density_functions.py

```diff
@@ -0,0 +1,1155 @@
+            d1, d2 -> dimensions of the Hilbert spaces of the chosen particles
+    Output: the observable Mana. Caution: it is only defined for d1,d2 = 3, 5
+    """
+    if (d1 not in [3, 5]) or (d2 not in [3, 5]):
+          raise ValueError("Mana is only defined for odd-dimension Hilbert spaces. Please use it for adapted systems.")
+    return np.log2(Sum_Discrete_Wigner(rho, d1, d2))
+
+def PeresHorodecki_criterion(rho: list[complex, complex], pdg_code: list[int], epsilon=1e-7) -> tuple[bool, list[float]]:
+    """
+    Input:  rho -> density matrix
+            pdg_code -> pdg codes of the two particles in the density matrix
+            epsilon -> treshold under which the eigenvalues are taken as 0 to deal with numerical instabilities
+    Output: flag_entanglement -> True if the criterion is satisfied and shows entanglement
+            eigvals -> eigenvalues of the partially transposed density matrix
+    """
+    rhoTB = Partial_Transpose(rho, 2, pdg_code)
+    eigvals, _ = la.eigh(rhoTB)
+    for i in range(len(eigvals)):
+        if abs(eigvals[i]) < epsilon:
+            eigvals[i] = 0
+    flag_entanglement = False
+    for eig in eigvals:
+        if eig < 0.:
+            flag_entanglement = True
+            break
+
+    return flag_entanglement, eigvals
+
+def trace_distance(Matrix1: list[complex, complex], Matrix2: list[complex, complex]) -> float:
+    """
+       Input: Matrix1, Matrix2 -> 2 square matrices
+       Output: trace distance between the two matrices
+    """
+    aux1 = Matrix1 - Matrix2
+    aux2 = np.dot(np.conjugate(np.transpose(aux1)), aux1)
+    eigvals, eigvecs = la.eigh(np.array(aux2))
+    for o in range(len(eigvals)): #This is necessary because numerical errors can cause a almost zero eigenvalue to be negative.
+                if eigvals[o]/max(eigvals) < 1e-20:
+                        eigvals[o] = 0. 
+    
+    sqrt_rho = np.dot(eigvecs, np.dot(np.sqrt(np.diag(eigvals)), la.inv(eigvecs)))
+
+    return (np.trace(sqrt_rho).real)/2. #we only keep the real part
+
+def Fidelity(rho1: list[complex, complex], rho2: list[complex, complex]) -> float:
+    """
+       Input: rho1, rho2 -> density matrices
+       Output: fidelity of the two density matrices
+    """
+    eigvals1, eigvecs1 = la.eigh(np.array(rho1))
+    sqrt_rho1 = np.dot(eigvecs1, np.dot(np.sqrt(np.diag(eigvals1)), la.inv(eigvecs1)))
+    aux = np.dot(sqrt_rho1, np.dot(rho2, sqrt_rho1)) #F = Tr[sqrt(aux)]
+    eigvals2, _ = la.eigh(np.array(aux))
+
+    return np.sum(np.sqrt(eigvals2)) #the trace is the sum of the eigenvalues
+
+    
+def Fidelity_distance(Matrix1: list[complex, complex], Matrix2: list[complex, complex]) -> float:
+    """
+       Input: Matrix1, Matrix2 -> 2 square matrices
+       Output: fidelity distance between the two matrices
+    """
+    fidelity = Fidelity(Matrix1, Matrix2)
+    return np.sqrt(1 - fidelity**2)
+
+
+#########################END QI OBSERVABLES ############################
```

Comment:
Here my only comment is why not using object oriented programming for such function.
Most of them are taking rho / Matrix entry.
So defining them within the density class would likely make it easier to find/handle in the future no ?